python/pyarrow/_dataset.pyx

```diff
@@ -42,6 +43,51 @@ def _forbid_instantiation(klass, subclasses_instead=True):
     raise TypeError(msg)
 
 
+ctypedef CResult[shared_ptr[CRandomAccessFile]] CCustomOpen()
+
+cdef class FileSource:
+
+    cdef:
+        # XXX why is shared_ptr necessary here? CFileSource shouldn't need it
```

Comment:
It seems stack allocated members require `__cinit__` for some reason https://github.com/apache/arrow/pull/7156#discussion_r425330758

Reply:
The Cython object attribute isn't stack-allocated. My point is that, if it's segfaulting, it's perhaps because default initialization of FileSource, then assignment, is faulty?

Reply:
Well I'm still assigning to it in `__cinit__` so I'd be surprised. I'll check in c++ as well